app/models/sample.rb

```diff
@@ -328,6 +337,62 @@ def initiate_s3_cp
     LogUtil.log_err_and_airbrake("Failed to upload S3 sample '#{name}' (#{id}): #{e}")
   end
 
+  def initiate_basespace_file_transfer
+    # Only continue if the sample status is CREATED and not yet UPLOADED.
+    return unless status == STATUS_CREATED
+
+    Rails.logger.info("HERE #{basespace_dataset_id} #{basespace_access_token}")
+
+    files = files_for_basespace_dataset(basespace_dataset_id, basespace_access_token)
+
+    raise "Could not fetch Basespace files for dataset #{basespace_dataset_id} for sample '#{name}' (#{id})" if files.blank?
+
+    files.each do |file|
+      max_tries = 3
+      try = 0
+      time_between_tries = 10
+
+      Rails.logger.info("Starting upload of sample '#{name}' (#{id}) file '#{file[:name]}' from Basespace")
+      while try < max_tries
+        success = download_from_basespace_to_s3(file[:download_path], sample_input_s3_path, file[:name])
+
+        if success
+          break
+        else
+          Rails.logger.error("Try ##{try}: Upload of sample '#{name}' (#{id}) file '#{file[:name]}' from Basespace failed")
+          try += 1
+          if try < max_tries
+            Kernel.sleep(time_between_tries)
+          end
+        end
+      end
+
+      raise "Upload of sample '#{name}' (#{id}) file '#{file[:name]} from Basespace failed after #{max_tries} retries" unless try < max_tries
+
+      input_file = InputFile.new(
+        name: file[:name],
+        source_type: InputFile::SOURCE_TYPE_BASESPACE,
+        source: file[:source_path]
+      )
+
+      input_files << input_file
+    end
+
+    self.status = STATUS_UPLOADED
+    save
+  rescue => e
+    LogUtil.log_err_and_airbrake("SampleUploadFailedEvent: #{e}")
+
+    pr = new_failed_pipeline_run(self)
+    pr.error_message = PipelineRun::ERROR_BASESPACE_UPLOAD_FAILED
+    pr.save!
+    self.status = STATUS_CHECKED
+    save
```

Comment:
removing the `self.save` here would save one extra db call (since the `update` below is ensured

Reply:
Sounds good! I didn't realize `update` updated all hanged values, not just the value passed in `update`.

Reply:
oh nvm it doesn't. you could do `self.status = STATUS_CHECKED` and `self.basespace_access_token = nil` and `save` in the ensure